server/jetstream_cluster.go

```diff
@@ -1719,6 +1829,63 @@ func (js *jetStream) monitorStream(mset *stream, sa *streamAssignment, sendSnaps
 	}
 }
 
+// When we are migration denotes if we ourselves are part of the old peer set or the new one.
+// Both types will be running at the same time as we scale up to extend into the new cluster.
+// Once detected we will us our type to dictate our behavior.
+type peerMigrateType int8
+
+const (
+	oldPeerGroup = peerMigrateType(iota)
+	newPeerGroup
+)
+
+// Determine if we are migrating and if so if we are part of the old or new set.
+func (mset *stream) isMigrating() (bool, peerMigrateType) {
+	sa := mset.streamAssignment()
+	// During migration we will always be R>1, even when we start R1.
+	// So if we do not have a group or node we no we are not migrating.
+	if sa == nil || sa.Group == nil || sa.Group.node == nil {
+		return false, oldPeerGroup
+	}
+	// The sign of migration is if our group peer count != configured replica count.
+	if sa.Config.Replicas == len(sa.Group.Peers) {
+		return false, oldPeerGroup
+	}
+	// So we believe we are migrating here, need to determine if we are the old set or new set.
+	// We can shor circuit this based on our group assigned cluster vs our own.
+	if sa.Group.Cluster == mset.srv.cachedClusterName() {
+		return true, newPeerGroup
+	}
+	return true, oldPeerGroup
+}
+
+// Determine if we are migrating and if so if we are part of the old or new set.
+func (o *consumer) isMigrating() (bool, peerMigrateType) {
+	ca := o.consumerAssignment()
+	// During migration we will always be R>1, even when we start R1.
+	// So if we do not have a group or node we no we are not migrating.
+	if ca == nil || ca.Group == nil || ca.Group.node == nil {
```

Comment:
Same question regarding locking here, except that you would have to be more precise because below you are getting stream's lock.

Reply:
Actually, I think I am right since running your PR locally, got this race the very first time:
```
=== RUN   TestJetStreamClusterStreamAndConsumerScaleUpAndDown
==================
WARNING: DATA RACE
Read at 0x00c00104edd0 by goroutine 113:
  github.com/nats-io/nats-server/v2/server.(*consumer).isMigrating()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:1867 +0x8f
  github.com/nats-io/nats-server/v2/server.(*jetStream).monitorConsumer()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3413 +0xd11
  github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateConsumer.func1()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3140 +0x4c

Previous write at 0x00c00104edd0 by goroutine 248:
  github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateConsumer()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3124 +0xbc4
  github.com/nats-io/nats-server/v2/server.(*jetStream).processConsumerAssignment()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:2931 +0xb93
  github.com/nats-io/nats-server/v2/server.(*jetStream).applyMetaEntries()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:1361 +0x384
  github.com/nats-io/nats-server/v2/server.(*jetStream).monitorCluster()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:836 +0xb2d
  github.com/nats-io/nats-server/v2/server.(*jetStream).monitorCluster-fm()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:776 +0x39

Goroutine 113 (running) created at:
  github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/server.go:2995 +0x86
  github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateConsumer()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3140 +0xdad
  github.com/nats-io/nats-server/v2/server.(*jetStream).processConsumerAssignment()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:2931 +0xb93
  github.com/nats-io/nats-server/v2/server.(*jetStream).applyMetaEntries()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:1361 +0x384
  github.com/nats-io/nats-server/v2/server.(*jetStream).monitorCluster()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:836 +0xb2d
  github.com/nats-io/nats-server/v2/server.(*jetStream).monitorCluster-fm()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:776 +0x39

Goroutine 248 (running) created at:
  github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/server.go:2995 +0x86
  github.com/nats-io/nats-server/v2/server.(*jetStream).setupMetaGroup()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:584 +0x1092
  github.com/nats-io/nats-server/v2/server.(*Server).enableJetStreamClustering()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:486 +0x20a
  github.com/nats-io/nats-server/v2/server.(*Server).enableJetStream()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream.go:398 +0x1168
  github.com/nats-io/nats-server/v2/server.(*Server).EnableJetStream()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/jetstream.go:204 +0x651
  github.com/nats-io/nats-server/v2/server.(*Server).Start()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/server.go:1728 +0x1804
  github.com/nats-io/nats-server/v2/server.RunServer·dwrap·4178()
      /Users/ivan/dev/go/src/github.com/nats-io/nats-server/server/server_test.go:90 +0x39
==================
    testing.go:1152: race detected during execution of test
```

Reply:
Same as above, but have not seen any race complaints about this access. Have fixed (or tried to fix) the ones I have seen. The expansion of the update stream popped a few.

Reply:
Will run a few locally myself and take a look. But its not the lock you suspect.

Reply:
ok see it locally as well.

Reply:
Yes, I don't know if it needs js or something else, but I can get this race quite easily.

Reply:
Maybe you need to make those isMigrating() a function of jetStream and pass a stream and consumer (2 different funcs) and inside you need to have js lock.. but you know better..

Reply:
Fixed.